src/ParseFile.js

```diff
@@ -292,6 +300,52 @@ class ParseFile {
     );
   }
 
+  /**
+   * Sets metadata to be saved with file object. Overwrites existing metadata
+   * @param {Object} metadata Key value pairs to be stored with file object
+   */
+  setMetadata(metadata: Object) {
```

Comment:
We should stick the the same convention as _name and _url by making these private. file.metadata() and file.tags() should be enough, metadata is usually immutable. If you want to keep these that fine too.

Reply:
I would propose these be promises that call the storage adapter to fetch the metadata and tags? If the file has not been saved we could return what is set in _metadata or _tags without making a call to the storage adapter. If so I would like to add this at a later time because it would require adding additional methods to the storage adapters.

Reply:
When you originally proposed this feature. I thought metadata would be stored in the DB. This would be way more useful and we would have more control. Is there a reason it’s not there? I don’t think it would be a breaking change.

Reply:
The reason it does no save to the database now is because files are currently stored in the DB as just a string. I agree this would be great to add to the database but it would require storing the files as objects instead (unless you have some other solution?). My original proposal included links to the S3 documentation showing how metadata and tags are stored with files in S3 directly. My idea behind building all of this (beforeSaveFile, afterSaveFile, metadata and tags) is so that I can A) query against my S3 and have fields to query against and B) Build a new collection for any files that are saved by using the afterSaveFile hook to create a new object, with the location of the file and any other data that I want to store with it. I can then add this new "File" object as a pointer to whatever object I want. Maybe in the future we can add a default "_File" collection that does this automatically?

Reply:
If you think the setTags and setMetadata could be misleading, I can change it so that those values can only be passed into the options object when saving the file? so something like:
```
const file = new Parse.File('some-file.txt');
await file.save({ metadata: {foo: 'bar'}, tags: {bar: 'foo'});
```

Reply:
You already have them in the constructor. I think that is enough.

Reply:
so remove the setters and getters?

Reply:
You can leave the getters. `file.metadata()` and `file.tags()`

Reply:
Remove the setters?

Reply:
Yes

Reply:
@dplewis if I remove the setters than you would have no way of modifying the metadata and tags on the beforeSaveFile hook.

Reply:
@dplewis I added some documentation with examples to show how this PR combined with my PR on parse-server will work together